src/upload/handler/upload.handler.ts

```diff
@@ -0,0 +1,146 @@
+import fs from 'fs'
+import multer from 'multer'
+import path from 'path'
+import Libp2p from 'libp2p'
+import config from 'config'
+
+import { subscribeForOffer } from '../../communication'
+import { UploadJobStatus } from '../../definitions'
+import { loggingFactory } from '../../logger'
+import { ProviderManager } from '../../providers'
+import UploadJob from '../model/upload.model'
+import UploadClient from '../model/clients.model'
+
+export const UPLOAD_FOLDER = 'uploads'
+
+const logger = loggingFactory('upload.handler')
+
+type UploadRouteHandler = (req: any, res: any) => Promise<void>
+
+async function unlinkFiles (files: any[]): Promise<void> {
+  for (const file of files) {
+    await fs.promises.unlink(file.path)
+    logger.info(`File ${file.filename} removed from fs`)
+  }
+}
+
+async function isClientAllowedToUpload (req: any): Promise<boolean> {
+  const ip = req.clientIp
+  logger.debug(`Client IP address = ${ip}`)
+  const client = await UploadClient.findOne({ where: { ip } })
+
+  return !client || client.uploads < config.get<number>('uploadLimitPerPeriod')
+}
+
+async function increaseClientUploadCounter (req: any): Promise<void> {
+  const ip = req.clientIp
+  const client = await UploadClient.findOne({ where: { ip } })
+
+  if (!client) {
+    await UploadClient.create({ ip, uploads: 1 })
+  } else {
+    client.uploads += 1
+    await client.save()
+  }
+}
+
+const storage = multer.diskStorage({
+  destination (req, file, cb) {
+    cb(null, path.join(process.cwd(), UPLOAD_FOLDER))
+  },
+  filename (req, file, cb) {
+    cb(null, `${Date.now()}-${file.originalname.split('.')[0]}.${file.mimetype.split('/')[1]}`)
+  }
+})
+
+const uploadMiddleware = multer({
+  storage,
+  limits: { fileSize: config.get<number>('fileSizeLimit') }
+}).array('files')
+
+export default function (storageProvider: ProviderManager, libp2p: Libp2p): UploadRouteHandler {
+  return async (req: any, res: any): Promise<void> => {
+    // Check if this IP is not DDOS
+    if (!await isClientAllowedToUpload(req)) {
+      return res.status(400).json({
+        error: 'Not allowed'
+      })
+    }
+
+    uploadMiddleware(req, res, async (err: any) => {
+      if (err) {
```

Comment:
I think this callback deserves decomposition. At least 2 parts can be identified: error handling and upload logic.

Reply:
done